### PR TITLE
feat: add ignoreUndefinedProperties option

### PR DESCRIPTION
**Pull request by schmidt-sebastian**
This adds an optional setting to ignore nested undefined values in all API calls that accept user data.

Fixes https://github.com/googleapis/nodejs-firestore/issues/1031

### CODE REVIEW

**Comment by nojvek** [comment]
This is great. Thanks. 

Pretty cool how battle tested such a minor feature is. 

To be honest I’m far in love with the nodejs sdk compared to the browser sdk. The browser sdk is quite shallow and a pain to use. Doesn’t have listCollectionIds and a bunch of other commands that are here. I hope someday this sdk will be usable in the browser too with the same api. The api is well thought out.

**Comment by schmidt-sebastian** [comment]
Just FYI: `listCollectionIds()` and `listDocuments()` require Service Account credentials and are not callable with the Firebase user accounts that the Web SDK uses for authentication. For this reason, listCollectionIds() is not available in the Web SDK. 

Thank you for your kind words.

File: dev/test/ignore-undefined.ts [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+              .startAt(undefined);
+          }).to.throw(
+            '"undefined" values are only ignored in object properties.'
+          );
+        }
+      );
+    });
+  });
+
+  describe('when setting is disabled', () => {
```
Comment (wu-hui) [inline dev/test/ignore-undefined.ts:232]: `Reject undefined fields when setting is disabled`?

File: types/firestore.d.ts [hunk 2]
```diff
@@ -143,6 +143,13 @@ declare namespace FirebaseFirestore {
      * when the client becomes idle. Defaults to 1.
      */
     maxIdleChannels?: number;
+
+    /**
+     * Whether to skip nested properties that are set to `undefined` during object
+     * serialization. If set to `false` or omitted, the SDK will throw an
+     * exception when it encounters values of type `undefined`.
```
Comment (wu-hui) [inline types/firestore.d.ts:150]: Do you think we should also document the effects (or non-effect) during deserialization?

**Comment by wu-hui** [comment]
Sorry for the delayed review!

File: dev/test/ignore-undefined.ts (hunk 1 above)
Comment (schmidt-sebastian) [reply dev/test/ignore-undefined.ts:232]: The full test names are: 
"rejects undefined values when setting is disabled in set()"
"rejects undefined values when setting is disabled in create()"
...

The name starts with the first "define" in line 158.

File: types/firestore.d.ts (hunk 2 above)
Comment (schmidt-sebastian) [reply types/firestore.d.ts:150]: Sounds good. Changed to:

```
  /**
   * Whether to skip nested properties that are set to `undefined` during object
   * serialization. If set to `true`, these properties will be skipped and are 
   * not be written to Firestore. If the setting is set `false` or omitted, 
   * the SDK will throw an exception when it encounters properties of type 
   * `undefined`.
   */
```

**Comment by kladnik** [comment]
@schmidt-sebastian Might be a silly question, but how do I enable this option? I am using `firebase-admin` and I cannot find a way to enable it, but I am getting: 
```
Error: Value for argument "data" is not a valid Firestore document. Cannot use "undefined" as a Firestore value (found in field "progress.test"). If you want to ignore undefined values, enable `ignoreUndefinedProperties`.
```
Which means that this feature should already be working.

**Comment by schmidt-sebastian** [comment]
@kladnik See the following:

```
const firestore  = admin.firestore();
firestore.settings({ignoreUndefinedProperties:true});
```

Make sure that you have installed the latest version of `@google-cloud/firestore`. You may have to remove your `package-lock.json` and your `node_modules` folder.

**Comment by kladnik** [comment]
@schmidt-sebastian Thanks! Might be good to add this example in the docs somewhere, so it will help others aswell.

**Comment by guness** [comment]
@schmidt-sebastian , totally agree with @kladnik .

I was really happy that you solved this after spending two hours how to fix the problem until realizing it is not working somehow.

then took one more hour to find out that solution actually requires deletion of those two.

**Pull request merged**